src/Builder/FrontendDialectTransformer.cpp

```diff
@@ -452,20 +542,39 @@ class FrontendGenImpl {
     //  * maintain a list of the defined graph
     llvm::SmallVector<mlir::Type, 4> arg_types;
 
+    // Get a list of function attributes - including names of inputs and outputs
+    llvm::SmallVector<mlir::NamedAttribute, 4> funcAttrs;
+    llvm::SmallVector<llvm::StringRef, 4> inputNames;
+    llvm::SmallVector<llvm::StringRef, 4> outputNames;
+
     // Import the input tensor types that are not constant and not initialized.
-    for (const auto &input : graph.input())
-      if (!initializedTensors.ContainKey(legalize_name(input.name())))
+    for (const auto &input : graph.input()) {
+      if (!initializedTensors.ContainKey(legalize_name(input.name()))) {
+        inputNames.push_back(input.name());
         arg_types.emplace_back(ImportInputTensorType(input));
+      }
+    }
+
+    for (const auto &output : graph.output()) {
+      outputNames.push_back(output.name());
+    }
+
+    funcAttrs.emplace_back(builder_.getNamedAttr(
+        "input_names", builder_.getStrArrayAttr(inputNames)));
+    funcAttrs.emplace_back(builder_.getNamedAttr(
+        "output_names", builder_.getStrArrayAttr(outputNames)));
 
     // Create the main function.
     auto funcType = builder_.getFunctionType(arg_types, {});
-    auto mainFunc =
-        mlir::FuncOp::create(UnknownLoc(), name, funcType, /* attrs = */ {});
+    auto mainFunc = mlir::FuncOp::create(UnknownLoc(), name, funcType,
+        /* attrs = */ llvm::makeArrayRef(funcAttrs));
 
     // Emit the entry point operation which specifies the number of user
     // inputs and outputs.
     auto entryPoint = mlir::ONNXEntryPointOp::create(UnknownLoc(), mainFunc,
-        /*numInputs=*/graph.input().size() - graph.initializer().size(),
+        /*numInputs=*/graph.input().size() < graph.initializer().size()
```

Comment:
Lots of our inputs had huge negatives in this attribute and I decided to fix it. I'd be more than happy to remove this change if it's unpopular :)

Reply:
How does the initializer have more elements than the number of inputs in your case? I'm assuming that's what you mean by "huge negatives".

Reply:
You're right - The initializer had things like weights, whereas the inputs just had the input tensors. This might be a difference in how the models are exported - training vs inference mode.

Reply:
I see, in the examples I've seen, everything is an input and a subset of those inputs have initializers which means they are constant. You have initializers which are not also inputs? Or you have multiple initializers for the same input?

Reply:
Yes - it seems like our weights are treated as initializers that are not also inputs.

Reply:
How are those weights used by your model then? Are they global?

Reply:
They get used inside the FuncOp that is created for the graph, or am I misunderstanding? They are just normal weights, I'm not sure why they are in the initializer and not in the inputs other than that they are constant at export time.

Reply:
Ah my fault, I didn't mean used, I meant more like consumed by your model. How are they passed to your model. I'm trying to understand whether there is something going on in the ONNX model itself where you use the weights inside the main function without passing them in the argument list (or something like that), i.e. they are just globals.

Reply:
ah - yeah they're essentially globals. We load them out-of-band with regards to ONNX

Reply:
Got it, in that case the actual fix would have to look a little but different.

Out of all the actual input arguments to the main function, you would have to exclude the ones which have an initializer from counting towards `numInputs`.  If in your case it happens to be that none of your actual input arguments have initializers (i.e. you can get away with returning `graph.input().size()`) then that's just a happy corner case but in general you need to support the case where you have a mix. To do that you'd have to find which initializers initialize one of the input args and only subtract that number from the total number of input args.

Reply:
Thank you for clarifying - I'll fix that :)

Reply:
I think the latest commit should be the actual fix as you described.

Reply:
Looks good now thanks for that! :)

Reply:
Happy to help! :)